coderd/insights_test.go

```diff
@@ -570,6 +580,810 @@ func TestTemplateInsights(t *testing.T) {
+					if defaultValue == "" {
+						defaultValue = option.value
+					}
+					options = append(options, &proto.RichParameterOption{
+						Name:  option.name,
+						Value: option.value,
+					})
+				}
+				parameters = append(parameters, &proto.RichParameter{
+					Name:         parameter.name,
+					DisplayName:  parameter.name,
+					Type:         "string",
+					Description:  parameter.description,
+					Options:      options,
+					DefaultValue: defaultValue,
+				})
+			}
+
+			// Prepare all workspace resources (agents and apps).
+			var (
+				createWorkspaces []func(uuid.UUID)
+				waitWorkspaces   []func()
+			)
+			var resources []*proto.Resource
+			for _, user := range users {
+				user := user
+				for _, workspace := range user.workspaces {
+					workspace := workspace
+
+					if workspace.template != template {
+						continue
+					}
+					authToken := uuid.New()
+					agentClient := agentsdk.New(client.URL)
+					agentClient.SetSessionToken(authToken.String())
+					workspace.agentClient = agentClient
+
+					var apps []*proto.App
+					for _, app := range workspace.apps {
+						apps = append(apps, &proto.App{
+							Slug:         app.name,
+							DisplayName:  app.name,
+							Icon:         app.icon,
+							SharingLevel: proto.AppSharingLevel_OWNER,
+							Url:          "http://",
+						})
+					}
+
+					resources = append(resources, &proto.Resource{
+						Name: "example",
+						Type: "aws_instance",
+						Agents: []*proto.Agent{{
+							Id:   uuid.NewString(), // Doesn't matter, not used in DB.
+							Name: "dev",
+							Auth: &proto.Agent_Token{
+								Token: authToken.String(),
+							},
+							Apps: apps,
+						}},
+					})
+
+					var buildParameters []codersdk.WorkspaceBuildParameter
+					for _, buildParameter := range workspace.buildParameters {
+						buildParameters = append(buildParameters, codersdk.WorkspaceBuildParameter{
+							Name:  buildParameter.templateParameter.name,
+							Value: buildParameter.value,
+						})
+					}
+
+					createWorkspaces = append(createWorkspaces, func(templateID uuid.UUID) {
+						// Create workspace using the users client.
+						createdWorkspace := coderdtest.CreateWorkspace(t, user.client, firstUser.OrganizationID, templateID, func(cwr *codersdk.CreateWorkspaceRequest) {
+							cwr.RichParameterValues = buildParameters
+						})
+						workspace.id = createdWorkspace.ID
+						waitWorkspaces = append(waitWorkspaces, func() {
+							coderdtest.AwaitWorkspaceBuildJob(t, user.client, createdWorkspace.LatestBuild.ID)
+							ctx := testutil.Context(t, testutil.WaitShort)
+							ws, err := user.client.Workspace(ctx, workspace.id)
+							require.NoError(t, err, "want no error getting workspace")
+
+							workspace.agentID = ws.LatestBuild.Resources[0].Agents[0].ID
+						})
+					})
+				}
+			}
+
+			// Create the template version and template.
+			version := coderdtest.CreateTemplateVersion(t, client, firstUser.OrganizationID, &echo.Responses{
```

Comment:
nit: maybe we should depend on `dbgen` as much as possible to shorten the test execution and prevent potential flakes

Reply:
I evaluated doing that, but we didn't have any methods for generating template/workspace build parameters. Perhaps it could be a future improvement?

Reply:
👍